src/main/java/com/github/iahrari/orderexample/config/PricingServiceHandler.java

```diff
@@ -0,0 +1,90 @@
+package com.github.iahrari.orderexample.config;
+
+import java.util.Optional;
+
+import com.github.iahrari.orderexample.dto.AuthResponse;
+import com.github.iahrari.orderexample.dto.AuthenticationUserDTO;
+import com.github.iahrari.orderexample.dto.PriceModelDTO;
+import com.github.iahrari.orderexample.exception.PriceResponseException;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestTemplate;
+
+import lombok.RequiredArgsConstructor;
+
+@Component
+@RequiredArgsConstructor
+public class PricingServiceHandler {
+    public final static String PRICE_ENDPOINT = "/api/v1/price";
+    private final static String AUTHENTICATION_ENDPOINT = "/api/v1/authentication";
+
+    private final RestTemplate restTemplate;
+
+    @Value("${pricing-service.username}")
+    private String username;
+    @Value("${pricing-service.password}")
+    private String password;
+    @Value("${pricing-service.url}")
+    private String pricingServiceUrl;
+
+    private volatile String token;
+
+    synchronized private void authenticate(){
+        try {
+            var user = AuthenticationUserDTO.builder()
+                            .username(username)
+                            .password(password)
+                            .build();
+            
+            var response = restTemplate.postForEntity(
+                                pricingServiceUrl + AUTHENTICATION_ENDPOINT, 
+                                user, AuthResponse.class
+                            );
+
+            token = response.getBody().getToken();
+        } catch (HttpClientErrorException e) {
+            throw new PriceResponseException(
+                        String.format(
+                            "Pricing service error code %d", 
+                            e.getStatusCode().value()
+                        ));
+        }
+    }
+
+    public Double getPrice(String source, String destination) {
+        var counter = 0;
+        var price = PriceModelDTO.builder()
+                .source(source)
+                .destination(destination)
+                .build();
+        
+        var headers = new HttpHeaders();
+        var entity = new HttpEntity<>(price, headers);
+
+        while(true){
```

Comment:
What is the purpose of an infinity loop?

Reply:
It's not actually an infinite loop. First this code tries to get price with token if the token is invalid or expired, then the catch clause will get a 401 error and it will try to get new token and because it's in loop after that again it tries to get price if this time it fails with 401 it will throw an exception otherwise it will return the price.